selfdrive/locationd/test/test_locationd.py

```diff
@@ -127,43 +130,53 @@ def send_msg(self, msg):
       waits_left -= 1
     time.sleep(0.0001)
 
-  def test_params_gps(self):
-    # first reset params
-    Params().put('LastGPSPosition', json.dumps({"latitude": 0.0, "longitude": 0.0, "altitude": 0.0}))
-
-    lat = 30 + (random.random() * 10.0)
-    lon = -70 + (random.random() * 10.0)
-    alt = 5 + (random.random() * 10.0)
-
-    for _ in range(1000):  # because of kalman filter, send often
-      msg = messaging.new_message('gpsLocationExternal')
-      msg.logMonoTime = 0
+  def get_fake_msg(self, name, t):
+    if name == "gpsLocationExternal":
+      msg = messaging.new_message(name)
       msg.gpsLocationExternal.flags = 1
       msg.gpsLocationExternal.verticalAccuracy = 1.0
       msg.gpsLocationExternal.speedAccuracy = 1.0
       msg.gpsLocationExternal.bearingAccuracyDeg = 1.0
       msg.gpsLocationExternal.vNED = [0.0, 0.0, 0.0]
-      msg.gpsLocationExternal.latitude = lat
-      msg.gpsLocationExternal.longitude = lon
-      msg.gpsLocationExternal.altitude = alt
-      self.send_msg(msg)
-
-    for _ in range(250):  # params is only written so often
-      msg = messaging.new_message('cameraOdometry')
-      msg.logMonoTime = 0
+      msg.gpsLocationExternal.latitude = self.lat
+      msg.gpsLocationExternal.longitude = self.lon
+      msg.gpsLocationExternal.altitude = self.alt
+    elif name == 'cameraOdometry':
+      msg = messaging.new_message(name)
       msg.cameraOdometry.rot = [0.0, 0.0, 0.0]
       msg.cameraOdometry.rotStd = [0.0, 0.0, 0.0]
       msg.cameraOdometry.trans = [0.0, 0.0, 0.0]
       msg.cameraOdometry.transStd = [0.0, 0.0, 0.0]
-      self.send_msg(msg)
+    elif name == 'sensorEvents':
+      msg = messaging.new_message(name, 0)
+    elif name in ['carState', 'liveCalibration']:
+      msg = messaging.new_message(name)
+    msg.logMonoTime = t
+    return msg
+
+  def test_params_gps(self):
+    # first reset params
+    Params().put('LastGPSPosition', json.dumps({"latitude": 0.0, "longitude": 0.0, "altitude": 0.0}))
 
+    self.lat = 30 + (random.random() * 10.0)
+    self.lon = -70 + (random.random() * 10.0)
+    self.alt = 5 + (random.random() * 10.0)
+    self.fake_duration = 90  # secs
+    # get fake messages at the correct frequency, listed in services.py
+    fake_msgs = [
+      self.get_fake_msg(name, int((i + j / services[name][1]) * 1e9)) for i in range(self.fake_duration)
+      for name in self.LLD_MSGS
+      for j in range(int(services[name][1]))
+    ]
+    for fake_msg in sorted(fake_msgs, key=lambda x: x.logMonoTime):
```

Comment:
why sort this? this isn't guaranteed at runtime

Reply:
Even when I publish messages in incorrect (unsorted logMonoTime) order?